File: app/src/main/java/net/veldor/flibustaloader/parsers/SearchResponseParser.java
package net.veldor.flibustaloader.parsers;

import android.util.Log;

import net.veldor.flibustaloader.App;
import net.veldor.flibustaloader.ui.OPDSActivity;

import org.w3c.dom.Document;
import org.w3c.dom.Node;
import org.w3c.dom.NodeList;
import org.xml.sax.InputSource;

import java.io.StringReader;
import java.util.ArrayList;

import javax.xml.parsers.DocumentBuilder;
import javax.xml.parsers.DocumentBuilderFactory;
import javax.xml.xpath.XPath;
import javax.xml.xpath.XPathConstants;
import javax.xml.xpath.XPathExpressionException;
import javax.xml.xpath.XPathFactory;

public class SearchResponseParser {

    private static final String BOOK_TYPE = "tag:book";
    private static final String GENRE_TYPE = "tag:root:genre";
    private static final String AUTHOR_TYPE = "tag:author";
    private static final String SEQUENCES_TYPE = "tag:sequences";
    private static final String SEQUENCE_TYPE = "tag:sequence";
    private static final CharSequence AUTHOR_SEQUENCE_TYPE = ":sequence:";
    private static final String NEW_GENRES = "tag:search:new:genres";
    private static final String NEW_SEQUENCES = "tag:search:new:sequence";
    private static final String NEW_AUTHORS = "tag:search:new:author";
    private NodeList mEntries;
    private XPath mXpath;

    public SearchResponseParser(String answer) throws XPathExpressionException {
        Document document = getDocument(answer);
        mXpath = XPathFactory.newInstance().newXPath();
        App.getInstance().mSearchTitle.postValue(((Node) mXpath.evaluate("/feed/title", document, XPathConstants.NODE)).getTextContent());
        mEntries = (NodeList) mXpath.evaluate("/feed/entry", document, XPathConstants.NODESET);
    }

    public ArrayList parseResponse() throws XPathExpressionException {
        // получу сущности
        if (mEntries != null && mEntries.getLength() > 0) {
            // в зависимости от типа сущностей запущу парсер
            int contentType = identificationSearchType(mEntries.item(0), mXpath);
            if (contentType > 0) {
                switch (contentType) {
                    case OPDSActivity.SEARCH_GENRE:
                        return GenresParser.parse(mEntries, mXpath);
                    case OPDSActivity.SEARCH_SEQUENCE:
                        return SequencesParser.parse(mEntries, mXpath);
                    case OPDSActivity
                            .SEARCH_AUTHORS:
                    case OPDSActivity
                            .SEARCH_NEW_AUTHORS:
                        return AuthorsParser.parse(mEntries, mXpath);
                    case OPDSActivity.SEARCH_BOOKS:
                        return BooksParser.parse(mEntries, mXpath);
                }
            }
        }
        return null;
    }

    private static Document getDocument(String rawText) {
        DocumentBuilderFactory dbFactory = DocumentBuilderFactory.newInstance();
        DocumentBuilder dBuilder;
        try {
            dBuilder = dbFactory.newDocumentBuilder();
            InputSource is = new InputSource(new StringReader(rawText));
            return dBuilder.parse(is);
        } catch (Exception e) {
            e.printStackTrace();
        }
        return null;
    }

    private int identificationSearchType(Node item, XPath xPath) throws XPathExpressionException {
        // получу идентификатор
        String id = ((Node) xPath.evaluate("./id", item, XPathConstants.NODE)).getTextContent();
        Log.d("surprise", "SearchResponseParser identificationSearchType 82: identification id is " + id);
        if (id.startsWith(BOOK_TYPE)) {
            return OPDSActivity.SEARCH_BOOKS;
        } else if (id.startsWith(GENRE_TYPE)) {
            return OPDSActivity.SEARCH_GENRE;
        } else if (id.startsWith(AUTHOR_TYPE)) {
            // проверю на возможность, что загружены серии
            if (id.contains(AUTHOR_SEQUENCE_TYPE)) {
                return OPDSActivity.SEARCH_SEQUENCE;
            } else {
                return OPDSActivity.SEARCH_AUTHORS;
            }
        } else if (id.startsWith(SEQUENCES_TYPE)) {
            return OPDSActivity.SEARCH_SEQUENCE;
        } else if (id.startsWith(SEQUENCE_TYPE)) {
            return OPDSActivity.SEARCH_SEQUENCE;
        } else if (id.startsWith(NEW_GENRES)) {
            return OPDSActivity.SEARCH_GENRE;
        } else if (id.startsWith(NEW_SEQUENCES)) {
            return OPDSActivity.SEARCH_SEQUENCE;
        } else if (id.startsWith(NEW_AUTHORS)) {
            return OPDSActivity.SEARCH_NEW_AUTHORS;
        }
        return -1;
    }

    public int getType() throws XPathExpressionException {
        if(mEntries.getLength() > 0){
            return identificationSearchType(mEntries.item(0), mXpath);
        }
        return -1;
    }
}

File: app/src/main/java/net/veldor/flibustaloader/parsers/SearchResponseParser.kt
package net.veldor.flibustaloader.parsers

import kotlin.Throws
import net.veldor.flibustaloader.App
import net.veldor.flibustaloader.ui.OPDSActivity
import org.w3c.dom.Document
import org.w3c.dom.Node
import org.w3c.dom.NodeList
import org.xml.sax.InputSource
import java.io.StringReader
import java.lang.Exception
import java.util.ArrayList
import javax.xml.parsers.DocumentBuilder
import javax.xml.parsers.DocumentBuilderFactory
import javax.xml.xpath.XPath
import javax.xml.xpath.XPathConstants
import javax.xml.xpath.XPathExpressionException
import javax.xml.xpath.XPathFactory

class SearchResponseParser(answer: String) {
    private val mEntries: NodeList?
    private val mXpath: XPath
    @Throws(XPathExpressionException::class)
    fun parseResponse(reservedSequenceName: String?): ArrayList<*>? {
        // получу сущности
        if (mEntries != null && mEntries.length > 0) {
            // в зависимости от типа сущностей запущу парсер
            val contentType = identificationSearchType(mEntries.item(0), mXpath)
            if (contentType > 0) {
                when (contentType) {
                    OPDSActivity.SEARCH_GENRE -> return GenresParser.parse(mEntries, mXpath)
                    OPDSActivity.SEARCH_SEQUENCE -> return SequencesParser.parse(mEntries, mXpath)
                    OPDSActivity.SEARCH_AUTHORS, OPDSActivity.SEARCH_NEW_AUTHORS -> return AuthorsParser.parse(
                        mEntries,
                        mXpath
                    )
                    OPDSActivity.SEARCH_BOOKS -> return BooksParser.parse(
                        mEntries,
                        mXpath,
                        reservedSequenceName
                    )
                }
            }
        }
        return null
    }

    @Throws(XPathExpressionException::class)
    private fun identificationSearchType(item: Node, xPath: XPath): Int {
        // получу идентификатор
        val id = (xPath.evaluate("./id", item, XPathConstants.NODE) as Node).textContent
        when {
            id.startsWith(BOOK_TYPE) -> {
                return OPDSActivity.SEARCH_BOOKS
            }
            id.startsWith(GENRE_TYPE) -> {
                return OPDSActivity.SEARCH_GENRE
            }
            id.startsWith(AUTHOR_TYPE) -> {
                // проверю на возможность, что загружены серии
                return if (id.contains(AUTHOR_SEQUENCE_TYPE)) {
                    OPDSActivity.SEARCH_SEQUENCE
                } else {
                    OPDSActivity.SEARCH_AUTHORS
                }
            }
            id.startsWith(SEQUENCES_TYPE) -> {
                return OPDSActivity.SEARCH_SEQUENCE
            }
            id.startsWith(SEQUENCE_TYPE) -> {
                return OPDSActivity.SEARCH_SEQUENCE
            }
            id.startsWith(NEW_GENRES) -> {
                return OPDSActivity.SEARCH_GENRE
            }
            id.startsWith(NEW_SEQUENCES) -> {
                return OPDSActivity.SEARCH_SEQUENCE
            }
            id.startsWith(NEW_AUTHORS) -> {
                return OPDSActivity.SEARCH_NEW_AUTHORS
            }
            else -> return -1
        }
    }

    @get:Throws(XPathExpressionException::class)
    val type: Int
        get() = if (mEntries!!.length > 0) {
            identificationSearchType(mEntries.item(0), mXpath)
        } else -1

    companion object {
        private const val BOOK_TYPE = "tag:book"
        private const val GENRE_TYPE = "tag:root:genre"
        private const val AUTHOR_TYPE = "tag:author"
        private const val SEQUENCES_TYPE = "tag:sequences"
        private const val SEQUENCE_TYPE = "tag:sequence"
        private val AUTHOR_SEQUENCE_TYPE: CharSequence = ":sequence:"
        private const val NEW_GENRES = "tag:search:new:genres"
        private const val NEW_SEQUENCES = "tag:search:new:sequence"
        private const val NEW_AUTHORS = "tag:search:new:author"
        private fun getDocument(rawText: String): Document? {
            val dbFactory = DocumentBuilderFactory.newInstance()
            val dBuilder: DocumentBuilder
            try {
                dBuilder = dbFactory.newDocumentBuilder()
                val `is` = InputSource(StringReader(rawText))
                return dBuilder.parse(`is`)
            } catch (e: Exception) {
                e.printStackTrace()
            }
            return null
        }
    }

    init {
        val document = getDocument(answer)
        mXpath = XPathFactory.newInstance().newXPath()
        App.instance.mSearchTitle.postValue(
            (mXpath.evaluate(
                "/feed/title",
                document,
                XPathConstants.NODE
            ) as Node).textContent
        )
        mEntries = mXpath.evaluate("/feed/entry", document, XPathConstants.NODESET) as NodeList
    }
}